dotnet/src/SemanticKernel.Core/Kernel.cs

```diff
@@ -209,9 +209,16 @@ public async Task<KernelResult> RunAsync(ContextVariables variables, Cancellatio
 
                 context = functionResult.Context;
 
+                var functionInvokedArgs = this.OnFunctionInvoked(functionDetails, functionResult);
+
+                if (functionInvokedArgs is not null)
```

Comment:
I'd expect to only create a new function result when needed e.g. the function invoked hook may have done nothing

```
If (functionInvokedArgs is not null && functionInvokedArgs.HasNewResult())
{
    functionResult = functionInvokedArgs.GetNewResult();
}
```